docs/madr/decisions/050-mesh-passthrough.md

```diff
@@ -0,0 +1,137 @@
+* **port**: port on which service can communicate
+* **protocol**: defines a protocol of the communication. Possible values:
+  * `Tls`: should be used when TLS traffic is originated by the client application in the case the `kuma.io/protocol` would be tcp
+  * `Tcp`: WARNING: can't be used when `match.type == Domain` (at TCP level we are not able to disinguish domain, in this case it is going to hijack whole traffic on this port). This will be validated in the config.
+  * `Grpc`
+  * `Http`
+  * `Http2`
+
+#### Implementation
+
+We can use [Matcher API](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/matching/matching_api.html#matching-api) which got out of a Alpha phase and is stable. We can use it to support different entries on one listener. Together with [passthrough](https://kuma.io/docs/2.7.x/networking/non-mesh-traffic/#outgoing) mode one the mesh, MeshOperator can disable all outgoing traffic except the one provided by `MeshPassthrough`. It is worth pointing out that when `passthrough.enabled` is set to `true` on the `Mesh`, `MeshPassthrough` policies have no effect because all outgoing traffic is allowed.
+
+From the envoy configuration point of view, we are going to add filter chain matchers under `outbound:passthrough:ipv4"` listener.
+
+#### Universal without transparent proxy
+
+This policy won't apply without transparent proxy.
+
+#### ZoneEgress
+
+TBA
+
+#### Security
+
+It is advised that the MeshOperator is responsible for the `MeshPassthrough` policy. This policy can introduce traffic outside of the mesh or even the cluster, and the MeshOperator should be aware of this. It is also important to note that once a policy is matched by tags, any service owner can add specific tags that match the policy. Depending on the security requirements, the MeshOperator:
+
+* control how service owners use these tags (OPA, Kyverno, MeshConstraints...),
+* should conduct audits and proceed based on the audit results.
+
+### Other options
+
+#### MeshExternalService resource with passthrough mode support
+
+As described in `MeshExternalService`(MADR-xxx TBA), we can create a separate `type: Passthrough`. 
+
+```yaml
+kind: MeshExternalService
+metadata:
+  name: example
+  namespace: kuma-system
+  labels:
+    kuma.io/mesh: default
+    kuma.io/zone: east-1
+spec:
+  match:
+  - type: Domain
+    value: *.aws.us-east-2.com
+    port: 80
+    protocol: http
+  - type: Domain
+    value: httpbin.com
+    port: 80
+    protocol: http
+  - type: CIDR
+    value: 10.1.1.0/24
+    port: 80
+    protocol: http
+  - type: IP
+    value: 192.168.0.1
+    port: 80
+    protocol: http
+  type: Passthrough
+```
+
+### Positive Consequences
+
+* One policy handling both cases
+
+### Negative Consequences
+
+* Confusing way of targeting `MeshExternalServices` with policies but not affecting `Passthrough` clusters
+* Match with many entries doesn't fit logically with `type: Managed`
+* Not possible to apply only on specific sidecar
```

Comment:
Does all matches go to one passthrough cluster or do we have separate clusters?
Or to rephrase, do we have separate stats or not?

Reply:
as mentioned on slack we have 3 options:
1. one cluster for all policies
2. cluster per policy (with multiple matches)
3. always cluster per match
4. additional field to specify stats name and only then generate separate cluster

1. is of least value
2. is problematic because we don't have that information (maybe we could use inspect API).
3. is a bit problematic because we'd have to use "type" + "value" + "port" + "protocol" - not easiest to read but can be that way
4. mix of 1 and 3

----

Let's start with 1. and go with 4. if we need to

Reply:
When you say:

> one cluster for all policies

You mean for all `MeshPassthrough` policy?

Reply:
What would be the envoy cluster name and stat prefix here? Same again the Envoy XDS config should become a bit more "official APIs"

Reply:
> You mean for all MeshPassthrough policy?

yes

> What would be the envoy cluster name and stat prefix here?

In which case? Let me try to come up with one for all of them:

1. meshpassthrough
2. meshpassthrough_{policyName}
3. meshpassthrough_{type}_{value}_{port}_{protocol}
4. meshpassthrough_{userSuppliedName} and `meshpassthrough_{type}_{value}_{port}_{protocol}` for the ones that don't have `userSuppliedName`